File: src/builtins/realpath.cpp
// Implementation of the realpath builtin.
#include "config.h"  // IWYU pragma: keep

#include "realpath.h"

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cwchar>

#include "../builtin.h"
#include "../common.h"
#include "../fallback.h"  // IWYU pragma: keep
#include "../io.h"
#include "../parser.h"
#include "../path.h"
#include "../wcstringutil.h"
#include "../wgetopt.h"
#include "../wutil.h"  // IWYU pragma: keep

struct realpath_cmd_opts_t {
    bool print_help = false;
    bool no_symlinks = false;
};

static const wchar_t *const short_options = L"+:hs";
static const struct woption long_options[] = {
    {L"no-symlinks", no_argument, nullptr, 's'}, {L"help", no_argument, nullptr, 'h'}, {}};

static int parse_cmd_opts(realpath_cmd_opts_t &opts, int *optind,  //!OCLINT(high ncss method)
                          int argc, const wchar_t **argv, parser_t &parser, io_streams_t &streams) {
    const wchar_t *cmd = argv[0];
    int opt;
    wgetopter_t w;
    while ((opt = w.wgetopt_long(argc, argv, short_options, long_options, nullptr)) != -1) {
        switch (opt) {
            case 's': {
                opts.no_symlinks = true;
                break;
            }
            case 'h': {
                opts.print_help = true;
                break;
            }
            case ':': {
                builtin_missing_argument(parser, streams, cmd, argv[w.woptind - 1]);
                return STATUS_INVALID_ARGS;
            }
            case '?': {
                builtin_unknown_option(parser, streams, cmd, argv[w.woptind - 1]);
                return STATUS_INVALID_ARGS;
            }
            default: {
                DIE("unexpected retval from wgetopt_long");
            }
        }
    }

    *optind = w.woptind;
    return STATUS_CMD_OK;
}
/// An implementation of the external realpath command. Doesn't support any options.
/// In general scripts shouldn't invoke this directly. They should just use `realpath` which
/// will fallback to this builtin if an external command cannot be found.
maybe_t<int> builtin_realpath(parser_t &parser, io_streams_t &streams, const wchar_t **argv) {
    const wchar_t *cmd = argv[0];
    realpath_cmd_opts_t opts;
    int argc = builtin_count_args(argv);
    int optind;

    int retval = parse_cmd_opts(opts, &optind, argc, argv, parser, streams);
    if (retval != STATUS_CMD_OK) return retval;

    if (opts.print_help) {
        builtin_print_help(parser, streams, cmd);
        return STATUS_CMD_OK;
    }

    if (optind + 1 != argc) {  // TODO: allow arbitrary args. `realpath *` should print many paths
        streams.err.append_format(BUILTIN_ERR_ARG_COUNT1, cmd, 1, argc - optind);
        builtin_print_help(parser, streams, cmd);
        return STATUS_INVALID_ARGS;
    }

    const wchar_t *arg = argv[optind];

    if (!opts.no_symlinks) {
        if (auto real_path = wrealpath(arg)) {
            streams.out.append(*real_path);
        } else {
            if (errno) {
                // realpath() just couldn't do it. Report the error and make it clear
                // this is an error from our builtin, not the system's realpath.
                streams.err.append_format(L"builtin %ls: %ls: %s\n", cmd, arg,
                                          std::strerror(errno));
            } else {
                // Who knows. Probably a bug in our wrealpath() implementation.
                streams.err.append_format(_(L"builtin %ls: Invalid arg: %ls\n"), cmd, arg);
            }

            return STATUS_CMD_ERROR;
        }
    } else {
        // We need to get the *physical* pwd here.
        auto realpwd = wrealpath(parser.vars().get_pwd_slash());
        if (!realpwd) {
            streams.err.append_format(L"builtin %ls: realpath failed: %s\n", cmd, std::strerror(errno));
            return STATUS_CMD_ERROR;
        }
        wcstring absolute_arg = string_prefixes_string(L"/", arg) ? arg : path_apply_working_directory(arg, *realpwd);
        streams.out.append(normalize_path(absolute_arg, /* allow leading double slashes */ false));
    }

    streams.out.append(L"\n");

    return STATUS_CMD_OK;
}

File: src/builtins/realpath.rs
//! Implementation of the realpath builtin.

use errno::errno;

use super::prelude::*;
use crate::env::Environment;
use crate::{
    path::path_apply_working_directory,
    wutil::{normalize_path, wrealpath},
};

#[derive(Default)]
struct Options {
    print_help: bool,
    no_symlinks: bool,
}

const short_options: &wstr = L!("+hs");
const long_options: &[WOption] = &[
    wopt(L!("no-symlinks"), NoArgument, 's'),
    wopt(L!("help"), NoArgument, 'h'),
];

fn parse_options(
    args: &mut [&wstr],
    parser: &Parser,
    streams: &mut IoStreams,
) -> Result<(Options, usize), ErrorCode> {
    let cmd = args[0];

    let mut opts = Options::default();

    let mut w = WGetopter::new(short_options, long_options, args);

    while let Some(c) = w.next_opt() {
        match c {
            's' => opts.no_symlinks = true,
            'h' => opts.print_help = true,
            ':' => {
                builtin_missing_argument(parser, streams, cmd, args[w.wopt_index - 1], false);
                return Err(STATUS_INVALID_ARGS);
            }
            ';' => {
                builtin_unexpected_argument(parser, streams, cmd, args[w.wopt_index - 1], false);
                return Err(STATUS_INVALID_ARGS);
            }
            '?' => {
                builtin_unknown_option(parser, streams, cmd, args[w.wopt_index - 1], false);
                return Err(STATUS_INVALID_ARGS);
            }
            _ => panic!("unexpected retval from WGetopter"),
        }
    }

    Ok((opts, w.wopt_index))
}

/// An implementation of the external realpath command. Doesn't support any options.
/// In general scripts shouldn't invoke this directly. They should just use `realpath` which
/// will fallback to this builtin if an external command cannot be found.
pub fn realpath(parser: &Parser, streams: &mut IoStreams, args: &mut [&wstr]) -> BuiltinResult {
    let cmd = args[0];
    let (opts, optind) = parse_options(args, parser, streams)?;

    if opts.print_help {
        builtin_print_help(parser, streams, cmd);
        return Ok(SUCCESS);
    }

    if optind == args.len() {
        builtin_print_help(parser, streams, cmd);
        return Err(STATUS_INVALID_ARGS);
    }

    let mut had_error = false;

    for &arg in &args[optind..] {
        if !opts.no_symlinks {
            if let Some(real_path) = wrealpath(arg) {
                streams.out.append(&real_path);
                streams.out.append(L!("\n"));
            } else {
                let errno = errno();
                if errno.0 != 0 {
                    streams.err.appendln(&wgettext_fmt!(
                        "builtin %s: %s: %s",
                        cmd,
                        arg,
                        errno.to_string()
                    ));
                } else {
                    streams
                        .err
                        .appendln(&wgettext_fmt!("builtin %s: Invalid arg: %s", cmd, arg));
                }
                had_error = true;
            }
        } else {
            let realpwd = wrealpath(&parser.vars().get_pwd_slash());

            if let Some(realpwd) = realpwd {
                let absolute_arg = if arg.starts_with(L!("/")) {
                    arg.to_owned()
                } else {
                    path_apply_working_directory(arg, &realpwd)
                };
                streams.out.append(&normalize_path(&absolute_arg, false));
                streams.out.append(L!("\n"));
            } else {
                streams.err.appendln(&wgettext_fmt!(
                    "builtin %s: realpath failed: %s",
                    cmd,
                    errno().to_string()
                ));
                had_error = true;
            }
        }
    }

    if had_error {
        Err(STATUS_CMD_ERROR)
    } else {
        Ok(SUCCESS)
    }
}
